keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Poisson.java

```diff
@@ -78,27 +76,10 @@ private static int sample(double mu, KeanuRandom random) {
 
     @Override
     public DoubleTensor logProb(IntegerTensor k) {
-        Tensor.FlattenedView<Double> muFlattenedView = mu.getFlattenedView();
-        Tensor.FlattenedView<Integer> kFlattenedView = k.getFlattenedView();
-
-        double[] result = new double[(int) k.getLength()];
-        for (int i = 0; i < result.length; i++) {
-            result[i] = Math.log(pmf(muFlattenedView.getOrScalar(i), kFlattenedView.getOrScalar(i)));
-        }
 
-        return DoubleTensor.create(result, k.getShape());
-    }
+        DoubleTensor kDouble = k.toDouble();
+        DoubleTensor logFactorialK = kDouble.plus(1).logGammaInPlace();
 
-    private static double pmf(double mu, int k) {
-        if (k >= 0 && k < 20) {
-            return (Math.pow(mu, k) / factorial(k)) * Math.exp(-mu);
-        } else if (k >= 20) {
-            double sumOfFactorial = 0;
-            for (int i = 1; i <= k; i++) {
-                sumOfFactorial += Math.log(i);
-            }
-            return Math.exp((k * Math.log(mu)) - sumOfFactorial - mu);
-        }
-        return 0;
+        return kDouble.times(mu.log()).minusInPlace(mu).minusInPlace(logFactorialK);
```

Comment:
Math looks good! (was the flattened array used before for efficiency?)

Reply:
The flattened array approach would be less efficient actually. I think we just didn't think to use the gamma function instead for factorial.